File: iOS/CleanVsDirtyCode/Dirty Code/DirtyCodeService.swift
//
//  DirtyCodeService.swift
//  CleanVsDirtyCode
//
//  Created by Alex Albu on 09.04.2023.
//

import Foundation

enum ShapeType: Int {
    case square = 0
    case rectangle
    case triangle
    case circle
}

struct ShapeUnion {
    let `type`: ShapeType
    let height: Double
    let width: Double
}


class DirtyCodeService {
    private let CTable: [Double] = [1.0, 1.0, 0.5, .pi]
    
    init() {
        
    }
    
    func getTimeForTotalArea() -> ContinuousClock.Instant.Duration {
        let clock = ContinuousClock()
        let result = clock.measure {
            var total: Double = 0
            for shapeUnion in generatedShapeUnions() {
                total += getAreaUnion(shape: shapeUnion)
            }
        }
        print("\(result) for dirty code implementation")
        return result
    }
    
    private func getAreaUnion(shape: ShapeUnion) -> Double {
        return CTable[shape.type.rawValue] * shape.width * shape.height
    }
    
    private func generatedShapeUnions() -> [ShapeUnion] {
        var shapeUnions: [ShapeUnion] = []
        for _ in 1...100000 {
            let type = ShapeType(rawValue: Int.random(in: 0..<4))!
            let height = Double.random(in: 1...10)
            let width = Double.random(in: 1...10)
            let shapeUnion = ShapeUnion(type: type, height: height, width: width)
            shapeUnions.append(shapeUnion)
        }
        return shapeUnions
    }
    
}

File: Ionic/src/app/services/dirty-code.service.ts
import { Injectable } from "@angular/core";

enum ShapeType {
    Square = 0,
    Rectangle,
    Triangle,
    Circle
}

interface ShapeUnion {
    type: ShapeType;
    height: number;
    width: number;
}

@Injectable({
    providedIn: 'root'
})
export class DirtyCodeService {

    private CTable: number[] = [1.0, 1.0, 0.5, Math.PI];

    public getTimeForTotalArea(): number {
        const start = Date.now();
        let total = 0;
        for (const shape of this.generatedShapeUnions()) {
            total += this.getAreaUnion(shape);
        }
        const end = Date.now();
        // Time elapsed in seconds
        return (end - start) / 1000;
    }

    private getAreaUnion(shape: ShapeUnion): number {
        return this.CTable[shape.type] * shape.height * shape.width;
    }

    private generatedShapeUnions(): ShapeUnion[] {
        let shapeUnions: ShapeUnion[] = [];
        for (let i = 1; i <= 100000; i++) {
            const type: ShapeType = Math.floor(Math.random() * 4);
            const height: number = Math.random() * 10 + 1;
            const width: number = Math.random() * 10 + 1;
            const shapeUnion: ShapeUnion = { type, height, width };
            shapeUnions.push(shapeUnion);
        }
        return shapeUnions;
    }
}
